pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -521,45 +534,81 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 	// Note that we do not resurrect Pods in this interval. Also note that scaling will take precedence over
 	// updates.
 	for target := len(condemned) - 1; target >= 0; target-- {
+		target := target
 		// wait for terminating pods to expire
 		if isTerminating(condemned[target]) {
 			logger.V(4).Info("StatefulSet is waiting for Pod to Terminate prior to scale down",
 				"statefulSet", klog.KObj(set), "pod", klog.KObj(condemned[target]))
 			// block if we are in monotonic mode
 			if monotonic {
-				return &status, nil
+				break
 			}
 			continue
 		}
 		// if we are in monotonic mode and the condemned target is not the first unhealthy Pod block
 		if !isRunningAndReady(condemned[target]) && monotonic && condemned[target] != firstUnhealthyPod {
 			logger.V(4).Info("StatefulSet is waiting for Pod to be Running and Ready prior to scale down",
 				"statefulSet", klog.KObj(set), "pod", klog.KObj(firstUnhealthyPod))
-			return &status, nil
+			break
 		}
 		// if we are in monotonic mode and the condemned target is not the first unhealthy Pod, block.
 		if !isRunningAndAvailable(condemned[target], set.Spec.MinReadySeconds) && monotonic && condemned[target] != firstUnhealthyPod {
 			logger.V(4).Info("StatefulSet is waiting for Pod to be Available prior to scale down",
 				"statefulSet", klog.KObj(set), "pod", klog.KObj(firstUnhealthyPod))
-			return &status, nil
+			break
 		}
-		logger.V(2).Info("Pod of StatefulSet is terminating for scale down",
-			"statefulSet", klog.KObj(set), "pod", klog.KObj(condemned[target]))
 
-		if err := ssc.podControl.DeleteStatefulPod(set, condemned[target]); err != nil {
-			return &status, err
-		}
-		if getPodRevision(condemned[target]) == currentRevision.Name {
-			status.CurrentReplicas--
+		fns = append(fns, func() (bool, error) {
+			logger.V(2).Info("Pod of StatefulSet is terminating for scale down",
+				"statefulSet", klog.KObj(set), "pod", klog.KObj(condemned[target]))
+			return true, ssc.podControl.DeleteStatefulPod(set, condemned[target])
+		})
+
+		if monotonic {
+			break
 		}
-		if getPodRevision(condemned[target]) == updateRevision.Name {
-			status.UpdatedReplicas--
+	}
+
+	if monotonic {
+		for _, fn := range fns {
+			if shouldExit, err := fn(); shouldExit || err != nil {
+				return &status, err
+			}
 		}
-		if monotonic {
-			return &status, nil
+	} else {
+		// Parallelize with slow start.
```

Comment:
This code structure suggest we may want to have some helper function like:
func RunInBatches(fns, initialBatchSize).

I wouldn't be much surprised if such function already exists somewhere.

Reply:
Refactored. The function is based on the one from ReplicaSet controller, but it has to be slightly different because we use different actions here rather than run the same function a given number of times.